src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -3446,6 +3473,38 @@ public void visitTry(JCTry tree) {
             super.visitTry(tree);
         }
 
+        @Override
+        public void visitForLoop(JCForLoop tree) {
+
+            // Recurse on initialization, condition, and step
+            scan(tree.init);
+            scan(tree.cond);
+            scan(tree.step);
+
+            // Collect the for() loop iteration variables with the FOR_LOOP_BODY_MAY_CAPTURE flag
+            HashSet<VarSymbol> capturableForLoopVars = null;
+            if (tree.init != null) {
```

Comment:
This is a bit unfortunate. In principle, a variable has `FOR_LOOP_BODY_MAY_CAPTURE` should just be  `EFFECTIVELY_FINAL` - at least in the context of the for loop body. I suppose there's still a problem that we'd have to unset `EFFECTIVELY_FINAL` after the loop body... but I wonder, do we care? After all, we have already flow-analyzed the other for-loop parts (condition and step), so perhaps we just set `EFFCTIVELY_FINAL` before scanning the loop body, and leave it there?

Reply:
We can't really mess with `EFFECTIVELY_FINAL` at all, because we need to preserve the current behavior in the init, condition, and step, so that you still get an error in this example:
```java
for (int i = 0;
  i++ + ((IntSupplier)() -> i).getAsInt() < 3; ) {      // error - i is not effectively final
    ((IntSupplier)() -> i).getAsInt();                  // ok - i is effectively final in the body
}
```
Put another way, I don't see how we could avoid having two separate "effectively final" flags - one for the loop body, and the original one for everything else, at least, not without combining `AssignAnalyzer` and `CaptureAnalyzer`.

Reply:
True. The sequencing between the two visitors is problematic here.